File: April leetcoding challenge/Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        #locate the mid of the list
        endNode = midNode = head
        num = ""
        llLen = 0
        while endNode != None:
            llLen += 1
            midNode = midNode.next
            endNode = endNode.next
            if endNode:
                llLen += 1
                endNode = endNode.next
        #reverse the second half of the list in place
        prev = None
        cur = midNode
        while cur != None:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        #check if the first half and second half of the list are equal
        halfLen = llLen // 2
        while halfLen > 0:
            if head.val != prev.val:
                return False
            head = head.next
            prev = prev.next
            halfLen -= 1
        return True
        
            
            

File: April leetcoding challenge/Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* midNode = head;
        ListNode* endNode = head;
        int llLen = 0;
        while(endNode != nullptr) {
            llLen++;
            midNode = midNode->next;
            endNode = endNode->next;
            if(endNode) {
                llLen++;
                endNode = endNode->next;
            }
        }
        ListNode* prev = nullptr;
        ListNode* cur = midNode;
        while(cur != nullptr) {
            ListNode* nxt = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nxt;
        }
        int halfLen = llLen/2;
        while(halfLen > 0) {
            if(head->val != prev->val) {
                return false;
            }
            head = head->next;
            prev = prev->next;
            halfLen--;
        }
        return true;
    }
};  
